File: monke_bars/corpus.py
"""Corpus construction: PostRecords -> a ranked, tiered pandas DataFrame.

Steps mirror the thesis method and are platform-agnostic:
  1. flatten PostRecords into a table
  2. drop duplicate posts (same post_id)
  3. optionally keep only posts published inside a date window
  4. optionally keep only one detected language
  5. score by engagement (likes + comments), rank, and split into equal tiers
The resulting DataFrame is what every analysis function consumes.
"""

from __future__ import annotations

import re
from typing import Iterable, Optional, Sequence

import pandas as pd
from langdetect import DetectorFactory, LangDetectException, detect

from .ingest.base import PostRecord

DetectorFactory.seed = 0  # deterministic language detection

_RE_HASHTAG = re.compile(r"#\w+")
_RE_MENTION = re.compile(r"@\w+")


def detect_language(caption_text: str) -> str:
    """ISO language code of a caption's prose (hashtags/mentions removed first)."""
    prose = _RE_MENTION.sub("", _RE_HASHTAG.sub("", caption_text or "")).strip()
    if len(prose) < 10:
        return "unknown"
    try:
        return detect(prose)
    except LangDetectException:
        return "unknown"


def assign_tiers(n_rows: int, labels: Sequence[str]) -> list[str]:
    """Assign each rank position (0-based, already sorted best-first) a tier label.

    The corpus is split into ``len(labels)`` equal bands by rank: the first band
    gets ``labels[0]``, and so on. Works for any number of tiers.
    """
    k = len(labels)
    band = n_rows / k if n_rows else 1
    out = []
    for i in range(n_rows):
        idx = min(int(i // band), k - 1)
        out.append(labels[idx])
    return out


def language_counts(df: pd.DataFrame) -> pd.DataFrame:
    """Which languages the corpus actually contains, commonest first.

    Interfaces should offer these, so nobody has to guess a code. The
    counts also expose detection failures: a large ``unknown`` share usually means
    short captions, and a language nobody expected usually means a lookalike.
    """
    if df.empty or "language" not in df.columns:
        return pd.DataFrame(columns=["Language", "Posts"])
    counts = df["language"].value_counts()
    return pd.DataFrame({"Language": counts.index, "Posts": counts.values})


def filter_languages(df: pd.DataFrame, language) -> pd.DataFrame:
    """Keep posts in one language or several, without re-running detection.

    Detection is by far the slowest step in the pipeline, so an interface that
    rebuilt the corpus on every checkbox change would be unusable. Detect once
    into the ``language`` column, then filter with this as often as needed.
    """
    if language is None or df.empty or "language" not in df.columns:
        return df.reset_index(drop=True) if not df.empty else df
    wanted = {language} if isinstance(language, str) else set(language)
    wanted = {str(w).lower() for w in wanted}
    return df[df["language"].str.lower().isin(wanted)].reset_index(drop=True)


def retier(df: pd.DataFrame, tier_labels: Sequence[str]) -> pd.DataFrame:
    """Re-rank and re-tier a corpus that has been filtered since it was built.

    Tiers are positions in a ranking, so any filter invalidates them: drop the
    top third of a corpus and the remaining posts still carry their old labels,
    which would put "High" on posts that are now the weakest in the set. An
    interface that filters interactively has to call this after every change.

    Kept separate from :func:`build_corpus` because language detection is the
    expensive step and must not be repeated. Detect once, then filter and retier
    as often as the controls demand.
    """
    if df.empty:
        return df
    out = df.sort_values("engagement_score", ascending=False, kind="stable").reset_index(drop=True)
    out["engagement_rank"] = out.index + 1
    out["engagement_tier"] = assign_tiers(len(out), list(tier_labels))
    return out


def filter_dates(df: pd.DataFrame, since=None, until=None) -> pd.DataFrame:
    """Keep posts published inside a window, without rebuilding the corpus.

    Posts with no timestamp are dropped whenever a bound is set, because there is
    no honest way to place them inside it.
    """
    if (since is None and until is None) or df.empty or "timestamp" not in df.columns:
        return df
    ts = pd.to_datetime(df["timestamp"], utc=True, errors="coerce")
    keep = ts.notna()
    if since is not None:
        keep &= ts >= pd.Timestamp(since, tz="UTC")
    if until is not None:
        keep &= ts <= pd.Timestamp(until, tz="UTC")
    return df[keep].reset_index(drop=True)


def build_corpus(
    posts: Iterable[PostRecord],
    language: Optional[str] = "en",
    tier_labels: Sequence[str] = ("High", "Medium", "Low"),
    detect_lang: bool = True,
    since=None,
    until=None,
) -> pd.DataFrame:
    """Build the analysis-ready corpus DataFrame from PostRecords.

    Parameters
    ----------
    posts : iterable of PostRecord
    language : str, list of str, or None
        Keep only posts detected as these languages. ``None`` keeps every post.
        A list is useful where one audience spans codes, for example Dutch and
        Flemish content that detection splits, or a market read in two languages.
    tier_labels : sequence of str
        Engagement tiers, best-first (default High/Medium/Low).
    detect_lang : bool
        Run language detection. Turned off internally when ``language`` is None
        and no ``language`` column is needed.
    since, until : date-like or None
        Keep only posts published in this window. Anything pandas can parse
        works, including ``"2025-08-01"``. Posts with no timestamp are dropped
        when a window is set, because there is no honest way to place them in
        it. The count of those is recorded on ``df.attrs["undated_removed"]``.
    """
    rows = [p.to_dict() for p in posts]
    df = pd.DataFrame(rows)
    if df.empty:
        return df

    # 1. dedup on post_id
    before = len(df)
    df = df.drop_duplicates(subset="post_id", keep="first").reset_index(drop=True)
    df.attrs["duplicates_removed"] = before - len(df)

    # 2. date window, before language detection so the slow step runs on fewer rows
    if since is not None or until is not None:
        ts = pd.to_datetime(df.get("timestamp"), utc=True, errors="coerce")
        keep = ts.notna()
        df.attrs["undated_removed"] = int((~keep).sum())
        if since is not None:
            keep &= ts >= pd.Timestamp(since, tz="UTC")
        if until is not None:
            keep &= ts <= pd.Timestamp(until, tz="UTC")
        before_date = len(df)
        df = df[keep].reset_index(drop=True)
        df.attrs["dated_out"] = before_date - len(df)
        if df.empty:
            return df

    # 3. language
    if detect_lang or language is not None:
        df["language"] = df["caption_text"].apply(detect_language)
    if language is not None:
        df = filter_languages(df, language)

    if df.empty:
        return df

    # 4. engagement score, rank, tier
    df["engagement_score"] = df["engagement_score"].fillna(0).astype(int)
    df = df.sort_values("engagement_score", ascending=False, kind="stable").reset_index(drop=True)
    df["engagement_rank"] = df.index + 1
    df["engagement_tier"] = assign_tiers(len(df), list(tier_labels))

    # 5. tidy hashtag/mention columns back into lists for convenience
    df["hashtag_list"] = df["hashtags"].apply(
        lambda s: [t.strip() for t in s.split(",") if t.strip()] if isinstance(s, str) else []
    )
    df["mention_list"] = df["mentions"].apply(
        lambda s: [t.strip() for t in s.split(",") if t.strip()] if isinstance(s, str) else []
    )
    df["hashtag_count"] = df["hashtag_list"].apply(len)
    return df

File: web/js/corpus.js
// Corpus construction: posts into a ranked, tiered, filterable set.
//
// A port of monke_bars/corpus.py. The expensive step, language detection, runs
// once in buildCorpus. Everything after it (filterLanguages, filterDates,
// retier) is cheap slicing, which is the whole reason this runs in the browser:
// a filter change touches an array already in memory and nothing else.

/** Tier labels for positions 0..n-1, already sorted best first. Equal bands. */
export function assignTiers(n, labels) {
  const k = labels.length;
  const band = n ? n / k : 1;
  const out = new Array(n);
  for (let i = 0; i < n; i++) out[i] = labels[Math.min(Math.floor(i / band), k - 1)];
  return out;
}

/** Rank by engagement and cut into tiers. A stable sort keeps tied posts in
 *  input order, matching the Python package: nearly half a capture can tie. */
export function retier(posts, labels) {
  const out = posts.slice().sort((a, b) => b.engagement_score - a.engagement_score);
  const tiers = assignTiers(out.length, labels);
  return out.map((p, i) => ({ ...p, engagement_rank: i + 1, engagement_tier: tiers[i] }));
}

/** Languages actually present, commonest first. Ties keep first appearance,
 *  as pandas value_counts does. */
export function languageCounts(posts) {
  const counts = new Map();
  for (const p of posts) counts.set(p.language, (counts.get(p.language) || 0) + 1);
  return [...counts.entries()]
    .sort((a, b) => b[1] - a[1])
    .map(([language, posts]) => ({ language, posts }));
}

export function filterLanguages(posts, languages) {
  if (!languages || !languages.length) return posts.slice();
  const want = new Set(languages.map((l) => String(l).toLowerCase()));
  return posts.filter((p) => want.has(String(p.language).toLowerCase()));
}

/** Posts published inside a window. Posts without a timestamp are dropped
 *  whenever a bound is set, since there is no honest way to place them in it. */
export function filterDates(posts, since = null, until = null) {
  if (!since && !until) return posts.slice();
  const lo = since ? Date.parse(since) : null;
  // An "until" date means through the end of that day.
  const hi = until ? Date.parse(until) + (String(until).length <= 10 ? 86399999 : 0) : null;
  return posts.filter((p) => {
    if (!p.timestamp) return false;
    const t = Date.parse(p.timestamp);
    return (lo === null || t >= lo) && (hi === null || t <= hi);
  });
}

/**
 * Dedup, detect language once, and tier.
 *
 * `detect` is injected: the browser passes franc, and the tests pass the
 * Python package's own labels so that only the port is under test.
 */
export function buildCorpus(posts, { detect, tierLabels = ["High", "Medium", "Low"], onProgress } = {}) {
  const seen = new Set(), unique = [];
  for (const p of posts) { if (seen.has(p.post_id)) continue; seen.add(p.post_id); unique.push(p); }

  const labelled = unique.map((p, i) => {
    if (onProgress && i % 25 === 0) onProgress(i, unique.length);
    return { ...p, language: detect(p.caption_text, p) };
  });
  const ranked = retier(labelled, tierLabels);
  return {
    posts: ranked.map((p) => ({ ...p, hashtag_list: p.hashtags, mention_list: p.mentions,
                                hashtag_count: p.hashtags.length })),
    duplicatesRemoved: posts.length - unique.length,
  };
}

// --- language detection -------------------------------------------------------

const W = "\\p{L}\\p{N}_";
const STRIP_TAG = new RegExp(`#[${W}]+`, "gu");
const STRIP_MENTION = new RegExp(`@[${W}](?:[${W}.]*[${W}])?`, "gu");

// franc speaks ISO 639-3; the stopword lists and the rest of the tool use 639-1.
const ISO3_TO_1 = {
  eng: "en", nld: "nl", deu: "de", fra: "fr", spa: "es", por: "pt", ita: "it", afr: "af",
  dan: "da", nob: "no", nno: "no", swe: "sv", fin: "fi", pol: "pl", ind: "id", zlm: "ms",
  tur: "tr", rus: "ru", hun: "hu", ron: "ro", ces: "cs", slk: "sk", cym: "cy", swh: "sw",
  som: "so", tgl: "tl", vie: "vi", arb: "ar", est: "et", cat: "ca", slv: "sl", hrv: "hr",
  lit: "lt", lvs: "lv", ell: "el", ukr: "uk", heb: "he", hin: "hi", jpn: "ja", kor: "ko",
  cmn: "zh-cn", tha: "th", ben: "bn", tam: "ta", urd: "ur", kaz: "kk", uzn: "uz", tgk: "tg",
  sqi: "sq", eus: "eu", bel: "be", nep: "ne",
};

// Only languages the tool has stopword lists for. franc's full model covers
// several hundred, and small relatives act as attractors on short captions:
// measured against langdetect on 432 real posts, English came back as Scots
// eleven times and Dutch as Low German. Restricting the candidates removed
// both and raised agreement from 77 to 82 percent.
export const FRANC_ONLY = ["eng", "nld", "deu", "fra", "spa", "por", "ita", "afr", "dan", "nob",
  "swe", "fin", "ind", "tur", "rus", "hun", "ron", "slv", "cat", "ell", "heb", "arb", "ben",
  "tam", "kaz", "uzn", "tgk", "sqi", "eus", "bel", "cmn", "hin", "nep"];

/** Wrap franc into the detector buildCorpus expects. Hashtags and mentions are
 *  removed first and short remainders are left undetected, as in Python. */
export function makeDetector(franc) {
  return (caption) => {
    const prose = (caption || "").replace(STRIP_TAG, "").replace(STRIP_MENTION, "").trim();
    if (prose.length < 10) return "unknown";
    const code = franc(prose, { minLength: 10, only: FRANC_ONLY });
    return code === "und" ? "unknown" : (ISO3_TO_1[code] || code);
  };
}
